src/Tmds.Ssh/GssapiWithMicCredential.cs

```diff
@@ -0,0 +1,284 @@
+        }
+
+        return true;
+    }
+
+    private async Task<bool> TryStageOid(SshConnection connection, string userName, CancellationToken ct)
+    {
+        {
+            using var userAuthMsg = CreateOidRequestMessage(connection.SequencePool,
+                                        userName, KRB5_OID);
+            await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+        }
+
+        using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+        ReadOnlySequence<byte>? oidResponse = GetGssapiOidResponse(response);
+
+        if (oidResponse is null)
+        {
+            return false;
+        }
+
+        return KRB5_OID.AsSpan().SequenceEqual(
+            oidResponse.Value.IsSingleSegment ? oidResponse.Value.FirstSpan : oidResponse.Value.ToArray());
+    }
+
+    private async Task<bool> TryStageAuthentication(SshConnection connection, NegotiateAuthentication authContext, CancellationToken ct)
+    {
+        byte[]? outToken = authContext.GetOutgoingBlob(Array.Empty<byte>(), out var statusCode);
+        while (outToken is not null)
+        {
+            {
+                using var userAuthMsg = CreateGssapiTokenMessage(connection.SequencePool, outToken);
+                await connection.SendPacketAsync(userAuthMsg.Move(), ct).ConfigureAwait(false);
+            }
+
+            // If the context is complete we don't expect a response.
+            if (statusCode == NegotiateAuthenticationStatusCode.Completed)
+            {
+                break;
+            }
+
+            // If not complete we expect the response input token to continue the auth.
+            using Packet response = await connection.ReceivePacketAsync(ct).ConfigureAwait(false);
+            ReadOnlySequence<byte>? tokenResponse = GetGssapiTokenResponse(response);
+            if (tokenResponse is null)
+            {
+                break;
+            }
+
+            outToken = authContext.GetOutgoingBlob(
+                tokenResponse.Value.IsSingleSegment ? tokenResponse.Value.FirstSpan : tokenResponse.Value.ToArray(),
+                out statusCode);
+        }
+
+        return statusCode == NegotiateAuthenticationStatusCode.Completed;
+    }
+
+    private static Packet CreateOidRequestMessage(SequencePool sequencePool, string userName, ReadOnlySpan<byte> oid)
+    {
+        /*
+            byte      SSH_MSG_USERAUTH_REQUEST
+            string    user name (in ISO-10646 UTF-8 encoding)
+            string    service name (in US-ASCII)
+            string    "gssapi-with-mic" (US-ASCII method name)
+            uint32    n, the number of mechanism OIDs client supports
+            string[n] mechanism OIDs
+        */
+        using var packet = sequencePool.RentPacket();
+        var writer = packet.GetWriter();
+        writer.WriteMessageId(MessageId.SSH_MSG_USERAUTH_REQUEST);
+        writer.WriteString(userName);
```

Comment:
Is it permitted to leave out the realm part when sending the username?

Reply:
Yep, surprisingly the `username` portion in this step is the user sshd will map the authentication to. The SSH server verifies that this user is valid on the remote before even starting the Kerberos token exchange. This means that I can specify the target username as part of the SSH connection settings but use a completely different principal for authentication. On Linux as long as the Kerberos principal is mapped to that user using a `.k5login` file in the user's home dir it will log on as that user.

I added a test for this specific scenario https://github.com/tmds/Tmds.Ssh/pull/188/files#diff-ce2477da774fe35b103aac30b0d1ed4afd6e515ce3c5eb96ad791d09b4b2b27bR161. In this case both `testuser` and `testuser@REALM.TEST` is a valid user on the Linux target and both users have a `.k5login` file that map `testuser@REALM.TEST` to the local account allowing me to login as both users with the same Kerberos principal.

For Windows, the Windows host will automatically lookup the user in the realm it is joined to so this also works. I don't believe it has an analogue for mapping a Kerberos principal to any other account though so the username needs to match the Kerberos principal used.

Also as a side note, you can specify the connection username using a UPN string

```csharp
var settings = new SshClientSettings("server")
{
    UserName = "user@REALM.COM",
};
```

Reply:
> For Windows, the Windows host will automatically lookup the user in the realm it is joined to so this also works. 

This is what I was expecting/hoping.

> The SSH server verifies that this user is valid on the remote before even starting the Kerberos token exchange.

Why does this check pass when we include a realm? What user is then checked against?

> Also as a side note, you can specify the connection username using a UPN string right now, it's just not possible to do so in the connection string due to the current parsing logic.

For this PR, I'd like to require an non-empty username when creating the `Credential`. And with https://github.com/tmds/Tmds.Ssh/issues/189 we can improve how it works with `SshClientSettings`.

Reply:
There are two different usernames that come into play with Kerberos authentication.

+ The SSH username that is supplied as part of the SSH messages
+ The Kerberos principal that is used with Kerberos auth

The first one is supplied as part of the initial OID message and the signed MIC data and is the username that the SSH server will attempt to map to a user on that host. For Linux this is typically just a user with the same name (what we do in the tests) but you can also use services like SSSD and maybe a PAM extension to support things like looking up the user through LDAP. For now all the server does is validates that this user is a valid user.

The second one is provided as part of the Kerberos authentication process. This is the Kerberos user that you are going to authenticate with and is aligned to either the cached credential retrieved by kinit on *nix, your logged in user on Windows, or the explicitly provided username/password. The server will validate that these credentials are valid for the principal specified.

Once the server has verified that the client is the Kerberos principal it will then determine if that principal can run as that user. On Windows I believe this just means that the Kerberos principal is the user specified in the SSH message, on *nix this is done by checking if the SSH username contains a `.k5login` file that contains the Kerberos principal used in the authentication.

In this PR these two usernames are derived with this logic:

+ The SSH user is either the Kerberos principal `UserName` if provided and not null/empty with a fallback to the connection setting `UserName`
+ The Kerberos principal is either used from the explicitly provided `NetworkCredential` or a cached credential

In the test case the container has two Linux users on the SSH host, `testuser` and `testuser@REALM.TEST`. Both accounts have a `.k5login` file that maps the Kerberos principal `testuser@REALM.TEST`. This allows us to login as the Linux user `testuser` or `testuser@REALM.TEST` using the same Kerberos principal `testuser@REALM.TEST`.

> For this PR, I'd like to require an non-empty username when creating the Credential. And with https://github.com/tmds/Tmds.Ssh/issues/189 we can improve how it works with SshClientSettings.

We can certainly do this but I don't think it's a good idea. We need a way to use Kerberos with a cached credential which by definition is an empty username/password in the `credential`. In this case the `username` must have a valid so we need to use the `SshClientSettings.UserName` value so it's a problem that needs to be solved now. I think the approach that I have in this PR right now is a decent approach that aligns with the behaviour of the normal `ssh` binary where the SSH username is the explicitly provided value in the connection string (or `-o UserName=...` with a fallback to the current user) and it will use the cached Kerberos credential for Kerb auth which can be a separate user.

The only scenario not covered here is if you want to use a different SSH username when providing an explicit `credential`. This could be solved in the future using a separate string or bool option that says to favour the `SshClientSettings.UserName` over the explicitly provided `credential` but I don't think we need to solve that use case right now.

Reply:
Thanks for the clear explanation!

> Another option is to always treat the SshClientSettings.UserName as the SSH username provided and only use the explicit credential for the Kerberos auth exchange. Thinking about this more this might actually be a better option than what I have today as it supports all the scenarios I can think off.

:+1: I was thinking the same thing as I read through your text.

Reply:
Thanks for asking the right questions, always good to go through the details and figure out better ways of doing things!

Reply:
Totally agree!